File: xml/basic_script/update/xml_update.rb
require "rexml/document"

xml = REXML::Document.new File.new(ARGV[0])
doc = xml.root

#要素を追加
doc.add_element("data", {"id" => "3"}).add_element("details").add_text("added")

n = doc.get_elements("//data[@id='2']")[0]
n.add_attribute("type", "node")

#要素を削除
doc.elements["//data[@id='1']"].remove
#以下でも可
#doc.get_elements("//data[@id='1']")[0].remove
#doc.elements["data"].remove
#doc.delete_element("data")

#要素を置換
n.replace_child(n.elements[1], REXML::Element.new("text").add_text("update test"))

//要素の値を変更
n.elements[2].text = "after"

#属性の変更
n.attributes["ext"] = "updated"

puts doc

File: xml/basic_script/update/xml_update.scala
import scala.xml._
import scala.xml.transform.{RewriteRule, RuleTransformer}

val doc = XML.loadFile(args(0))

val rule = new RewriteRule {
	override def transform(n: Node): NodeSeq = n match {
		//要素の追加
		case <root>{ ch @ _* }</root> => <root>{ ch }<data id="3"><details>added</details></data></root>

		//要素の削除
		case e: Elem if (e \ "@id").text == "1" => NodeSeq.Empty

		//属性の追加と変更
		case e: Elem if (e \ "@id").text == "2" =>
			val ne = e % Attribute("", "type", "node", Null) % Attribute("", "ext", "updated", Null)
			//以下でも可
		//	e % new UnprefixedAttribute("type", "node", Null) % new UnprefixedAttribute("ext", "updated", Null)

			//子要素を処理
			val ch: NodeSeq = ne.child.zipWithIndex.map {l =>
				l._2 match {
					//要素を置換
					case 0 => <text>update test</text>
					//要素の値を変更
					case 1 => Elem(l._1.prefix, l._1.label, l._1.attributes, l._1.scope, Text("after"))
					case _ => l._1
				}
			}

			//子要素のみを変更した要素のコピーを作成
			ne.copy(child = ch)

		case n => n
	}

}

val newDoc = new RuleTransformer(rule).transform(Utility.trim(doc))

println(newDoc.mkString)
